frontends/p4/strengthReduction.cpp

```diff
@@ -444,10 +444,22 @@ const IR::Node* DoStrengthReduction::postorder(IR::Slice* expr) {
     }
 
     while (auto cast = expr->e0->to<IR::Cast>()) {
-        if (expr->getH() < size_t(cast->expr->type->width_bits())) {
-            expr->e0 = cast->expr;
+        if (auto tb = cast->expr->type->to<IR::Type_Bits>()) {
+            auto type = expr->type->to<IR::Type_Bits>();
+            CHECK_NULL(type);
+            if (tb->isSigned != type->isSigned)
```

Comment:
I always assumed that slices of signed bit strings will produce an unsigned bit string. This should probably ensure that a cast to `int<W>` is converted into a `bit<W>`. A cast from `int<W>` to bit<W> should be acceptable though? 

This is what the spec says: 
>The result is an unsigned bit-string of width m - l + 1, including the bits numbered from l (which becomes the least significant bit of the result) to m (the most significant bit of the result) from the source operand.

Reply:
I added another test case which fails without this change; it's in the test file:
`h.h.b = ((bit<1>)h.h.c)[0:0];`

Reply:
But
`((bit<1>)h.h.c)[0:0];`
to
`((bit<1>)h.h.c);` 
should be acceptable, right?

Reply:
The previous result was just h.h.c, which failed to typecheck.